extract_sets/training_sets.py

```diff
@@ -0,0 +1,398 @@
+        strings = self.element.findall(f'{self.namespace}:String', XML_NS)
+        self.text_words = [e.attrib['CONTENT'] for e in strings]
+
+    def get_shape(self, element):
+        x_1 = int(element.attrib['HPOS'])
+        y_1 = int(element.attrib['VPOS'])
+        y_2 = y_1 + int(element.attrib['HEIGHT'])
+        x_2 = x_1 + int(element.attrib['WIDTH'])
+        return Polygon([(x_1, y_1), (x_2, y_1), (x_2, y_2), (x_1, y_2)])
+
+    def get_next_element_height(self, element):
+        y_start = int(element.attrib['VPOS'])
+        y_height = int(element.attrib['HEIGHT'])
+        return y_start + y_height
+
+
+class PageLine(TextLine):
+    """Extract TextLine Information from PAGE Data"""
+
+    def __init__(self, element, namespace, reorder):
+        super().__init__(element, namespace)
+        self.set_id()
+        self.set_text()
+        if self.valid:
+            self.reorder = reorder
+            self.shape = self.get_shape(self.element)
+
+    def set_id(self):
+        self.element_id = self.element.attrib['id']
+
+    def set_text(self):
+        """
+        * set words as preferred text source, otherwise use text line
+        * drop rtl-mark if contained
+        * print lines without coords
+        """
+
+        texts = []
+        text_els = self.element.findall(f'{self.namespace}:Word', XML_NS)
+        for t in text_els:
+            top_left = PageLine._pick_top_left(t, self.namespace)
+            if not top_left:
+                elem_id = t.attrib['id']
+                msg = f"Invalid Coords of Word '{elem_id}' in '{self.element_id}'!"
+                raise RuntimeError(msg)
+            texts.append(t)
+
+        # if no Word assume at least TextLine exists
+        if not text_els:
+            top_left = PageLine._pick_top_left(self.element, self.namespace)
+            if not top_left:
+                elem_id = self.element.attrib['id']
+                print("[ERROR  ] skip '{}': invalid coords!".format(
+                    elem_id), file=sys.stderr)
+                self.valid = False
+                return
+            texts.append(self.element)
+
+        sorted_els = sorted(
+            texts,
+            key=lambda w: int(PageLine._pick_top_left(w, self.namespace)))
+        unicodes = [
+            w.find(
+                f'.//{self.namespace}:Unicode',
+                XML_NS) for w in sorted_els]
+        self.text_words = [u.text.strip() for u in unicodes if u.text]
+
+        # elimiate read order mark
+        for i, strip in enumerate(self.text_words):
+            strip = self.text_words[i]
+            for mark in CLEAR_MARKS:
+                if mark in strip:
+                    self.text_words[i] = strip.replace(mark, '')
+
+    @staticmethod
+    def _pick_top_left(elem, namespace):
+        coords = elem.find(f'{namespace}:Coords', XML_NS)
+        points = coords.attrib['points'].split()
+        if len(points) > 0:
+            return points[0].split(',')[0]
```

Comment:
That's not the top left – that's the first point pair's x coordinate (which could be anywhere in the polygon). 

It's probably not an issue if Words do not overlap, but why not just use the center, just to be on the safe side? 

Also, what about vertical text here? In that case you need to sort by y, not by x.

```suggestion
    def _pos(elem, namespace, vertical=False):
        coords = elem.find(f'{namespace}:Coords', XML_NS)
        points = coords.attrib['points'].split()
        center = tuple(map(lambda c: sum(c) / len(c), zip(*points)))
        if vertical:
            return center[1]
        return center[0]
```

Reply:
Sounds reasonable, I will do some tests this way

Reply:
What about calculating a bounding rect and then picking the polygon point with shortest distance? 
In case of a *real* rectangular shape, this should be the top left corner. And no need for vertical flag.

Reply:
> What about calculating a bounding rect and then picking the polygon point with shortest distance?
> In case of a _real_ rectangular shape, this should be the top left corner. 

Much too complicated. And Shapely's distance/project is not as easy to use as it looks on first sight (trust me). Above code does the job nicely. Plus you get the option for vertical sorting for free.

> And no need for vertical flag.

With vertical text you need to sort in y direction, not in x direction.